modules/core/shared/src/main/scala/io/circe/Printer.scala

```diff
@@ -118,89 +117,106 @@ final case class Printer(
-              }
-              encloseJsonString(key)
-              builder.append(p.colons)
-              trav(depth + 1, value)
-              first = false
-            }
+  private[this] final def printJsonAtDepth(writer: Appendable)(json: Json, depth: Int): Unit = {
+    if (json.isNull) writer.append(nullText) else (json: @unchecked) match {
+      case Json.JString(s) => printJsonString(writer)(s)
+      case Json.JNumber(n) => writer.append(n.toString)
+      case Json.JBoolean(b) => if (b) writer.append("true") else writer.append("false")
+      case Json.JObject(o) =>
+        val p = pieces(depth)
+        writer.append(p.lBraces)
+        val items = if (preserveOrder) o.toList else o.toMap
+        var first = true
+
+        val fieldIterator = items.iterator
+
+        while (fieldIterator.hasNext) {
+          val (key, value) = fieldIterator.next()
+          if (!dropNullKeys || !value.isNull) {
+            if (!first) writer.append(p.objectCommas)
+            printJsonString(writer)(key)
+            writer.append(p.colons)
+            printJsonAtDepth(writer)(value, depth + 1)
+            first = false
           }
-          builder.append(p.rBraces)
-        case Json.JString(s) => encloseJsonString(s)
-        case Json.JNumber(n) => builder.append(n.toString)
-        case Json.JBoolean(b) => builder.append(if (b) trueText else falseText)
-        case Json.JArray(arr) =>
-          val arrIterator = arr.iterator
-          if (!arrIterator.hasNext) builder.append(p.lrEmptyBrackets) else {
-            builder.append(p.lBrackets)
-            trav(depth + 1, arrIterator.next)
-
-            while (arrIterator.hasNext) {
-              builder.append(p.arrayCommas)
-              trav(depth + 1, arrIterator.next())
-            }
-            builder.append(p.rBrackets)
+        }
+        writer.append(p.rBraces)
+      case Json.JArray(a) =>
+        val p = pieces(depth)
+        val len = a.length
+
+        if (len == 0) writer.append(p.lrEmptyBrackets) else {
+          writer.append(p.lBrackets)
+          printJsonAtDepth(writer)(a(0), depth + 1)
+
+          var i = 1
+
+          while (i < len) {
+            writer.append(p.arrayCommas)
+            printJsonAtDepth(writer)(a(i), depth + 1)
+            i += 1
           }
-        case Json.JNull => builder.append(nullText)
-      }
+
+          writer.append(p.rBrackets)
+        }
     }
+  }
 
-    trav(0, j)
-    builder.toString
+  /**
+   * Returns a string representation of a pretty-printed JSON value.
+   */
+  final def pretty(json: Json): String = {
+    val writer = new StringBuilder()
+
+    printJsonAtDepth(writer)(json, 0)
+
+    writer.toString
+  }
+
+  final def prettyBytes(json: Json): Array[Byte] = {
+    val bytes = new ByteArrayOutputStream()
+    val writer = new BufferedWriter(new OutputStreamWriter(bytes, "UTF-8"))
+
+    printJsonAtDepth(writer)(json, 0)
+
+    writer.close()
+    bytes.toByteArray
```

Comment:
I wish we could avoid [copying here](https://docs.oracle.com/javase/7/docs/api/java/io/ByteArrayOutputStream.html#toByteArray()). Even though there is no API to retrieve the underlying buffer directly, people seem to use [this simple trick](http://stackoverflow.com/a/15686667) to do that. Please, note that this means we'd need to change the return type to `ByteBuffer` since the underlying `Array[Byte]` may contain more bytes than we need so we have to crop it.

I think we can also expose that as a standalone method (`printByteBuffer`) if we're worried about this extra complexity. What do you think @travisbrown?